### ISSUE TITLE
[Bug] Can't change user password

### ISSUE THREAD

**Issue by pavlov80** [DirectoryTree/LdapRecord#152]
**Environment (please complete the following information):**
 - LDAP Server Type: [ActiveDirectory]
 - PHP Version: [7.4]
 - LdapRecord-laravel [1.5.0]

**Describe the bug:**
Hi Steve,
I tried to change and to reset the password as written in the documentation:

Change password:
```php
$user = User::find('cn=John Doe,ou=Users,dc=local,dc=com');

$user->unicodepwd = ['old-password', 'new-password'];
```

Reset password:
```php
$user = User::find('cn=John Doe,ou=Users,dc=local,dc=com');

$user->unicodepwd = 'new-password';
```

Both return this error from AD:
`20Type or value exists0000207E: AtrErr: DSID-031906F6, #1: 0: 0000207E: DSID-031906F6, problem 1006 (ATT_OR_VALUE_EXISTS), data 0, Att 9005a (unicodePwd)`

Some more information:
- I use domain user with privileges for create and modify users
- I can create sucessfully new users
- I can sucessfully change other attributes than password
- The password meets the conditions for complexity
- I can change the password with the same user when I use adldap2 library, which we use at the moment
- I tried all possible variants of old and/or new password with or without 'UTF-16LE' encoding. Same error

**Comment by stevebauman** [comment]
Hi @pavlov80,

Can you run a `composer update` and try again to make sure you're on the most current version of LdapRecord?

If it still occurs, are you using SSL or TLS to connect to your server?

Also:

> I use domain user with privileges for create and modify users

This user must be a domain administrator to reset passwords. Is this the case?

I'm able to reset or modify passwords using TLS or SSL using a domain administrator account -- or bind as the user whom needs their password changed and update it using their old and new password.

**Comment by stevebauman** [comment]
Also -- is there more to your posted code? Or is the password change the only operation you're performing on the user account?

**Comment by pavlov80** [comment]
Yes, I am on the most curent version. My connection is using SSL on port 636. The user is domain administrator and I can reset the password with adldap2 library and same user. Yes, the password change is the only operation that I try to perform. If I change other attributes like mail, it's working as expected. 
One more thing, we use Multi-Domain Authentication.

The actual code that I use is this:

```
$user = Employee::where('samaccountname', '=', auth()->user()->username)->first();
$connection = $user->getConnection();
if ($connection->auth()->attempt($user->getDn(), $request->input('old_password'))) {
        // Create Unicode password. Assumes that given password is in UTF-8 encoding! Adjust it to the actual encoding of the password
	$pwdtxt = $request->input('new_password');
	$pwdtxt = '"' . $pwdtxt . '"';
	$newPass = iconv( 'UTF-8', 'UTF-16LE', $pwdtxt );

	$user->unicodepwd = $newPass;
       try {
		// User password changed!
		$user->save();
		return "OK";
			 
	} catch (LdapRecordException $ex) {
		// Failed changing password.
		$connection = $user->getConnection();

		// Get the last LDAP error to determine the cause of failure.
		$error = $connection->getLdapConnection()->getDetailedError();

		echo $error->getErrorCode();
		echo $error->getErrorMessage();
		echo $error->getDiagnosticMessage();
	}
}
```

**Comment by stevebauman** [comment]
Ok I see the issue here -- you are hashing the password twice. LdapRecord already does this for you via the `getHashedPassword()` method when you set the `unicodepwd` attribute:

https://github.com/DirectoryTree/LdapRecord/blob/974a0ac6669ec54e4401d2bfe9d76f3b0eda02d7/src/Models/Concerns/HasPassword.php#L38-L47

Set the `unicodepwd` attribute as the raw string and let me know if successful 👍

**Comment by stevebauman** [comment]
Also, if this user you are binding as is not a domain administrator, you must pass in their old password to be able to update it:

```php
if ($connection->auth()->attempt($user->getDn(), $request->input('old_password'))) {
    $user->unicodepwd = [$request->input('new_password'), $request->input('old_password')];

    $user->save();
}
```

**Comment by pavlov80** [comment]
I cannot create the user if I skip the convert thing on the password. I thought it was necessary in case of change too. Anyway, I tried with raw string and the error is the same.
Then I tried to bind with my user, on which I want to change the password. I pass the old, and new password and again this same error appears. I have no idea why it doesn't work.

**Comment by stevebauman** [comment]
> I cannot create the user if I skip the convert thing on the password.

Can you elaborate on this? Is this a new user that is being created in the same script? What happens when you skip converting the password? Need to know more here.

> The ActiveDirectory is on Windows Server 2012 I think.

Is this your own AD instance? Are you sure you've connected with a domain administrator account to test resetting passwords via:

```php
$user->unicodepwd = 'my-new-password';

$user->save();
```

**Comment by pavlov80** [comment]
I do not have any problems when **creating** the user if I convert the password first. Here is the **working** code on creating the user:

```
$fn = $fno->fn;
$email = $fn.'@domain.bg';
$cn = $fno->fullname;
$dn = 'cn='.$cn.' '.$fn.',ou=Test,dc=students,dc=domain,dc=bg';
$password = $this->convertToAdPassword($fno->password_plain);

$student = new Student;
$student->cn = $cn.' '.$fn;
$student->name = $cn.' '.$fn;
$student->displayName = $cn;
$student->givenname = $fno->firstname;
$student->sn = $fno->lastname;
$student->initials = $fno->initials;
$student->co = 'Bulgaria';
$student->c = 'BG';
$student->samaccountname = $fn;
$student->unicodepwd = $password;
$student->company = 'My Company';
$student->physicalDeliveryOfficeName = $fno->fakultet;
$student->department = $fno->katedra;
$student->mail = $email;
$student->UserPrincipalName = $email;
$student->description = $fno->description;
$student->UserAccountControl = 66048;
$student->title = 'Student';
$student->setDn($dn);
$student->save();
```

This is the code for **changing** the password and it is **not working**: I put only the method which is not working as I have many more code in this controller for other things.
```
<?php
namespace App\Http\Controllers\portal;

use App\Http\Controllers\Controller;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Cookie;
use App\Requests\portal\PasswordRequest;
use Session;
use DB;
use Illuminate\Support\Facades\Auth;
use App\Ldap\Student;
use App\Ldap\Employee;
use App\Ldap\Teacher;
use LdapRecord\LdapRecordException;
use LdapRecord\Models\ActiveDirectory\User;

class MainController extends Controller
{
   
    public function password_change_update(PasswordRequest $request)
    {
    	// Get the user instance
 		if(Auth::guard('employee')->check()) {
			$user = Employee::where('samaccountname', '=', auth()->user()->username)->first();
		} else {
			$user = Teacher::where('samaccountname', '=', auth()->user()->username)->first();
		}

		// Get the connection info and Distinguished Name
		$connection = $user->getConnection();

		// Check the password
		if ($connection->auth()->attempt($user->getDn(), $request->input('old_password'))) {

			$user->unicodepwd = 'Sup3rN0va.p4ssw0rd4';
			//$user->mail = 'my_user@domain.bg'; // this is working


			try {
				// User password changed!
			    $user->save();

			    return 'OK';
			 
			} catch (LdapRecordException $ex) {
			    // Failed changing password.
			    $connection = $user->getConnection();

			    // Get the last LDAP error to determine the cause of failure.
			    $error = $connection->getLdapConnection()->getDetailedError();

			    echo $error->getErrorCode();
			    echo $error->getErrorMessage();
			    echo $error->getDiagnosticMessage();
			}
		} else {
			return back()->withErrors(['the password do not match']);
		}
    }
}
```
I go to sleep now, but will see if you have any other suggestions tomorow. Thank you for your time!

**Comment by stevebauman** [comment]
Hi @pavlov80,

You're authenticating as the user and performing a password reset, not a change.

Can you change this:

```php
$user->unicodepwd = 'Sup3rN0va.p4ssw0rd4';
```

To:

```php
$user->unicodepwd = [$request->input('old-password'), 'Sup3rN0va.p4ssw0rd4'];
```

And see if that works?

**Comment by pavlov80** [comment]
I already tried all these combinations without success. The problem must be somewhere else.
In this file I just get instance of the user which I want to make changes.
The connections in my config/ldap.php are domain user with permission to change the passwords (tested). We are using certificate too.
In my /etc/ldap/ldap.conf I have this:
```
ssl start_tls
TLS_CACERT /etc/ssl/certs/my_cert.pem
TLS_REQCERT hard
```

**Comment by stevebauman** [comment]
Okay, I'm honestly not sure. I would try using SSL instead of TLS and giving it another go.

Other than that, you could try pushing a modification yourself to your LDAP server yourself and see if you encounter the same issue. It's working in my environment, so there is something I'm not seeing here.

I'll leave this issue open for others to assist you.

**Comment by pavlov80** [comment]
Good news, I resolved my issue.
I have Multi-Domain Authentication, and I make multiple models for every user type. In these files I extend this model:
`use LdapRecord\Models\Model;`
instead of:
`use LdapRecord\Models\ActiveDirectory\User;`

Now I can change my password successfully.
I'm glad I found these lines in the documentation that pointed me to the real problem.
> You may want these models to extend the built-in models, as they include functionality that you do not need to build yourself. It's completely up to you.

And thank you for this great library!
